Tasks API - Add task cancelation capability (#195)
meilisearch/specifications: text/0060-tasks-api.md

|------------|---------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| uid        | string | Unique sequential identifier                                                                                                                                                                                                  |
| indexUid   | string  | Unique index identifier. This field is `null` when the task is a [global task](#global-task).                                                   |
| status     | string  | Status of the task. Possible values are `enqueued`, `processing`, `succeeded`, `failed`, `canceled`                                                                                                                          |
| type       | string  | Type of the task. Possible values are `indexCreation`, `indexUpdate`, `indexDeletion`, `documentAdditionOrUpdate`, `documentDeletion`, `settingsUpdate`, `dumpCreation`, `taskCancelation`                                                    |

> | canceledBy | integer | Unique identifier of the `taskCancelation` task that canceled the given task. Default is set to `null`.                                                    |

Reviewer: ```suggestion
| canceledBy | integer | Unique identifier of the task that performed a `taskCancelation` on this task. Default is set to `null`.                                                    |
```

```suggestion
| canceledBy | integer | Unique identifier of the task that canceled this task. Default is set to `null`.                                                    |
```

| name    | description  |
| -----   | ------------ |

> | dumpUid | The generated uid of the dump. `null` when the task status is enqueued or processing. |

Reviewer 2: ```suggestion
| dumpUid | The generated uid of the dump. `null` when the task status is enqueued, processing, canceled or, failed. |
```

Reviewer 2: Signal to @maryamsulemani97. We are adding precision here that you may want to mention in the related API reference.

##### taskCancelation

| Name          | Description |
| ------------- | ----------- |
| matchedTasks  | The number of tasks that can be canceled based on the request. If the API key doesn’t have access to any of the indexes specified in the request via the `indexUids` query parameter, those tasks will not be included in `matchedTasks`. | 
| canceledTasks | The number of tasks successfully canceled. If the task fails, `0` is displayed. `null` when the task status is enqueued or processing. |

> | originalFilters | The extracted URL query parameters used in the originating task cancelation request. |

Reviewer 2: ```suggestion
| originalFilter | The extracted URL query parameters used in the originating task cancelation request. |
```

}
```

e.g. A fully qualified `task` object in a `canceled` state.

```json
{
    "uid": 0,
    "indexUid": "movies",
    "status": "canceled",
    "type": "settingsUpdate",
    "details": {
        "rankingRules": [
            "typo",
            "ranking:desc",
            "words",
            "proximity",
            "attribute",
            "exactness"
        ]
    },
    "canceledBy": 1,
    "enqueuedAt": "2021-08-10T14:29:17.000000Z",

>     "startedAt": "2021-08-10T14:29:18.000000Z",

Reviewer 2: I sense option `2` is better for the sake of consistency and task history ⚡

>     "startedAt": "2021-08-10T14:29:18.000000Z",

Reviewer 2: @Kerollmops, I missunderstood. I'm just saying that `startedAt` should always be displayed at `null` or with the exact time when it's started to be processed.

If a task has been debatched to be reprocess later, `startedAt` can go back to `null`

>     "startedAt": "2021-08-10T14:29:18.000000Z",

Reviewer 3: The behavior has been changed accordingly in the engine 👍

---

> ##### 6.2. Get a task by uid | `GET` - `/tasks/{uid}`

Reviewer: We should mention that there are query params and that they are the same as `GET /tasks`

`202` - Response body - `/tasks/cancel`

```json
{
    "taskUid": 0,
    "indexUid": null,
    "status": "enqueued",
    "type": "taskCancelation",
    "enqueuedAt": "2021-08-12T10:00:00.000000Z"
}
```

##### 6.3.2. Response Definition

When the request is successful, Meilisearch returns the HTTP code 202 Accepted. The response's content is the summarized representation of the received asynchronous task.

##### 6.3.3. Auto-batching

If the task you’re canceling is part of a batch, **the whole batch is stopped.** Meilisearch automatically creates a new batch once the current one is stopped and the specified tasks are canceled. The canceled tasks will not be part of the new batch.

This means:
- When the new batch is created, it may contain tasks that have been requested to be canceled in the meantime
- Any progress the batch made before being canceled is lost

##### 6.3.4. Errors

If a user tries canceling a `succeeded`, `failed`, or `canceled` task, it won’t throw an error. Task cancelation is an atomic transaction; all tasks are successfully canceled, or none aren't.

> - 🔴 Sending a task cancelation without filtering query parameters returns a `[missing_filters](https://github.com/meilisearch/specifications/blob/main/text/0061-error-format-and-definitions.md#missing_filters)` error.

Author: I agree it's more clear and accurate

---

##### 6.3. Cancel tasks | `POST` - `/tasks/cancel`

##### 6.3.1. Goals

Allows users to cancel an `enqueued` or `processing` task. Particularly useful if a long or heavy task blocks the queue.

`202` - Response body - `/tasks/cancel`

```json
{
    "taskUid": 0,
    "indexUid": null,
    "status": "enqueued",
    "type": "taskCancelation",
    "enqueuedAt": "2021-08-12T10:00:00.000000Z"
}
```

##### 6.3.2. Response Definition

When the request is successful, Meilisearch returns the HTTP code 202 Accepted. The response's content is the summarized representation of the received asynchronous task.

##### 6.3.3. Auto-batching

> If the task you’re canceling is part of a batch, **the whole batch is stopped.** Meilisearch automatically creates a new batch once the current one is stopped and the specified tasks are canceled. The canceled tasks will not be part of the new batch.

Reviewer: ```suggestion
If the task you’re canceling is part of a batch, **the whole batch is stopped.** Once it is stopped, Meilisearch automatically creates a new batch with the remaining tasks. The canceled tasks are not part of the new batch.
```

This means:
- When the new batch is created, it may contain tasks that have been enqueued between the batch cancelation and recreation.
- Any progress the batch made before being canceled is lost.

##### 6.3.4. Errors

> If a user tries canceling a `succeeded`, `failed`, or `canceled` task, it won’t throw an error. Task cancelation is an atomic transaction; all tasks are successfully canceled, or none aren't.

Reviewer 3: ```suggestion
If a user tries canceling a `succeeded`, `failed`, or `canceled` task, it won’t throw an error. Task cancelation is an atomic transaction; all tasks are successfully canceled, or none are.
```

###### 11.2.2. `indexUid`

- Type: String
- Required: False
- Default: `*`

`indexUid` is **case-sensitive**.

###### 11.2.3. `status`

- Type: String
- Required: False
- Default: `*`

- 🔴 If the `status` parameter value is not consistent with one of the task statuses, an [`invalid_task_status`](0061-error-format-and-definitions.md#invalidtaskstatus) error is returned.

###### 11.2.4. `type`

- Type: String
- Required: False
- Default: `*`

`type` is **case-insensitive**.

- 🔴 If the `type` parameter value is not consistent with one of the task types, an [`invalid_task_type`](0061-error-format-and-definitions.md#invalidtasktype) error is returned.

###### 11.2.5. Date Parameters

###### 11.2.5.1. `beforeEnqueuedAt` and `afterEnqueuedAt`

- Type: String
- Required: False
- Default: `*`

###### 11.2.5.2. `beforeStartedAt` and `afterStartedAt`

- Type: String
- Required: False
- Default: `*`

###### 11.2.5.3. `beforeFinishedAt` and `afterFinishedAt`

- Type: String
- Required: False
- Default: `*`

###### 11.2.5.4. Date Format

The filter accepts the RFC 3339 format. The following syntaxes are valid:

- `Y-M-D`
- `Y-M-DTH:M:SZ`

> - `Y-M-DTH:M:S+01:00`

Reviewer 3: Not sure about that. @loiclec can you confirm that we support those 3 formats, please?

Reviewer 4: There are two accepted syntaxes:
- `YYYY-MM-DD` (note the mandatory number of digits for year/month/day)
- a `date-time` defined by the "Internet Date/Time format" defined by RFC 3339 section 5.6: https://www.rfc-editor.org/rfc/rfc3339#section-5.6, which IMO we should not attempt to describe in detail in the specifications. But yes, in short it is something like `YYYY-MM-DDTHH-MM:SS` followed by an optional `.[0-9]+` followed by either `Z` or an offset `(+|-)HH:MM`

Reviewer 2: I've updated the spec to make it more valid regarding the accepted formats without going into details. A suggestion is welcome if you have ideas on how to describe this more precisely, as always!

| beforeStartedAt | string | No       | Permits to filter tasks by the time before they were started. By default, when `beforeStartedAt` or `afterStartedAt` query parameter is not set, all task types are concerned. It's possible to specify several types by separating them with the `,` character.                       |
| afterStartedAt | string | No       | Permits to filter tasks by the time after they were started. By default, when `beforeStartedAt` or `afterStartedAt` query parameter is not set, all task types are concerned. It's possible to specify several types by separating them with the `,` character.                       |
| beforeFinishedAt | string | No       | Permits to filter tasks by the time before they were finished. By default, when `beforeFinishedAt` or `afterFinishedAt` query parameter is not set, all task types are concerned. It's possible to specify several types by separating them with the `,` character.                       |

> | afterFinishedAt | string | No       | Permits to filter tasks by the time after they were finished. By default, when `beforeFinishedAt` or `afterFinishedAt` query parameter is not set, all task types are concerned. It's possible to specify several types by separating them with the `,` character.                       |

Reviewer 2: ```suggestion
| afterFinishedAt | string | No       | Filter tasks based on their finishedAt time. Retrieve tasks finished after the given filter value.                 |
```

###### 11.2.2. `indexUid`

- Type: String
- Required: False
- Default: `*`

`indexUid` is **case-sensitive**.

###### 11.2.3. `status`

- Type: String
- Required: False
- Default: `*`

- 🔴 If the `status` parameter value is not consistent with one of the task statuses, an [`invalid_task_status`](0061-error-format-and-definitions.md#invalid_task_status) error is returned.

###### 11.2.4. `type`

- Type: String
- Required: False
- Default: `*`

`type` is **case-insensitive**.

- 🔴 If the `type` parameter value is not consistent with one of the task types, an [`invalid_task_type`](0061-error-format-and-definitions.md#invalid_task_type) error is returned.

###### 11.2.5. `canceledBy`

- Type: Integer
- Required: False
- Default: `null`

`canceledBy` is **case-insensitive**.

- 🔴Sending a value with a different type than `Integer` for the `canceledBy` parameter returns an [`invalid_task_canceled_by`](0061-error-format-and-definitions.md#invalid_task_canceled_by) error.

###### 11.2.6. Date Parameters

Date filters accepts the RFC 3339 format. The following syntaxes are valid:

- `Y-M-D`
- `Y-M-DTH:M:SZ`
- `Y-M-DTH:M:S+01:00`

> ###### 11.2.6.1. `beforeEnqueuedAt` and `afterEnqueuedAt`

Reviewer 2: I agree!

- 🔴 If the `type` parameter value is not consistent with one of the task types, an [`invalid_task_type`](0061-error-format-and-definitions.md#invalidtasktype) error is returned.

###### 11.3.5. `beforeEnqueuedAt` and `afterEnqueuedAt`

- Type: String
- Required: False
- Default: `*`

The filter accepts the RFC 3339 format. The following syntaxes are valid:

- `Y-M-D`
- `Y-M-DTH:M:SZ`
- `Y-M-DTH:M:S+01:00`

> - 🔴 The date filters are exclusive. You can cancel tasks before or after a specified date, meaning it will not be included.

Author: I have changed the paragraphs tell me if it's better.
I also note that some titles have a capital letter for each word and others do not so I put Capital for new paragraph
